src/protocols/der/encode.c

```diff
@@ -0,0 +1,2014 @@
+{
+	fr_dbuff_t	our_dbuff = FR_DBUFF(dbuff);
+	uint8_t 	tag_byte;
+
+	tag_byte = (tag_class & DER_TAG_CLASS_MASK) | (constructed & DER_TAG_CONSTRUCTED_MASK) |
+		   (tag_num & DER_TAG_NUM_MASK);
+
+	fr_dbuff_in(&our_dbuff, tag_byte);
+
+	return fr_dbuff_set(dbuff, &our_dbuff);
+}
+
+/** Encode a DER structure
+ *
+ * @param[out] dbuff		The buffer to write the structure to
+ * @param[in] cursor	The cursor to the structure to encode
+ * @param[in] encode_ctx	The encoding context
+ *
+ * @return		The number of bytes written to the buffer
+ */
+static ssize_t encode_value(fr_dbuff_t *dbuff, UNUSED fr_da_stack_t *da_stack, UNUSED unsigned int depth,
+			    fr_dcursor_t *cursor, void *encode_ctx)
+{
+	fr_pair_t const	    *vp;
+	fr_dbuff_t	     our_dbuff = FR_DBUFF(dbuff);
+	fr_dbuff_marker_t    marker;
+	fr_der_tag_encode_t *tag_encode;
+	fr_der_tag_num_t     tag_num;
+	fr_der_tag_class_t   tag_class;
+	fr_der_encode_ctx_t *uctx = encode_ctx;
+	ssize_t		     slen = 0;
+
+	if (unlikely(cursor == NULL)) {
+		fr_strerror_const("No cursor to encode");
+		return -1;
+	}
+
+	vp = fr_dcursor_current(cursor);
+	if (unlikely(vp == NULL)) {
+		fr_strerror_const("No pair to encode");
+		return -1;
+	}
+
+	FR_PROTO_TRACE("Encoding %s", vp->da->name);
+
+	PAIR_VERIFY(vp);
+
+	/*
+	 *	ISO/IEC 8825-1:2021
+	 *	The structure of a DER encoding is as follows:
+	 *
+	 *		+------------+--------+-------+
+	 *		| IDENTIFIER | LENGTH | VALUE |
+	 *		+------------+--------+-------+
+	 *
+	 *	The IDENTIFIER is a tag that specifies the type of the value field and is encoded as follows:
+	 *
+	 *		  8   7    6    5   4   3   2   1
+	 *		+---+---+-----+---+---+---+---+---+
+	 *		| Class | P/C |     Tag Number    |
+	 *		+---+---+-----+---+---+---+---+---+
+	 *			   |
+	 *			   |- 0 = Primitive
+	 *			   |- 1 = Constructed
+	 *
+	 *	The CLASS field specifies the encoding class of the tag and may be one of the following values:
+	 *
+	 *		+------------------+-------+-------+
+	 *		|      Class       | Bit 8 | Bit 7 |
+	 *		+------------------+-------+-------+
+	 *		| UNIVERSAL        |   0   |   0   |
+	 *		| APPLICATION      |   0   |   1   |
+	 *		| CONTEXT-SPECIFIC |   1   |   0   |
+	 *		| PRIVATE          |   1   |   1   |
+	 *		+------------------+-------+-------+
+	 *
+	 *	The P/C field specifies whether the value field is primitive or constructed.
+	 *	The TAG NUMBER field specifies the tag number of the value field and is encoded as an unsigned binary
+	 *	integer.
+	 *
+	 *	The LENGTH field specifies the length of the VALUE field and is encoded as an unsigned binary integer
+	 *	and may be encoded as a single byte or multiple bytes.
+	 *
+	 *	The VALUE field contains LENGTH number of bytes and is encoded according to the tag.
+	 *
+	 */
+
+	if (fr_der_flag_has_default(vp->da)) {
+		/*
+		 *	Skip encoding the default value, as per ISO/IEC 8825-1:2021 11.5
+		 */
+		fr_dict_enum_value_t const *evp;
+
+		evp = fr_dict_enum_by_name(vp->da, "DEFAULT", strlen("DEFAULT"));
```

Comment:
I'll leave this in for now, but it could be in the `dict_ext` stuff?

Reply:
I made similar comments on the review call.  It can just be a flag.